datalake: add support for value_subject_latest iceberg_mode

rockwotj opened the pull request:
Add support for `value_subject_latest`.

This enables iceberg for users who don't use schema registry, but use protobuf and the backwards compatible nature of the wire format. Users will still have to register schemas with the schema registry, but data does not have be encoded seperately.

By default we assume that protobuf messages use the first message defined in the file descriptor and the subject is <topic>-value (matches SubjectNameStrategy). However both of these can be overridden using key=value syntax within the topic property. The syntax for this new iceberg mode something like: `value_subject_latest(:(<key>=<value>,)+)?` where the only defined keys are `subject` and `protobuf_name`.

Overall this PR has a rather small surface area:

The `iceberg_mode` enum is converted to a `struct`, but keeps a similar API to the enum so very little code needs to change. Additionally, the serde wire format is entirely backwards compatible.

We plug in a new implementation of the `type_resolver` interface for this behavior, then everything else is pretty straightforward.

## Backports Required

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

### Features

* Iceberg topics can be configured with `redpanda.iceberg.mode=value_subject_latest:protobuf_name=<full_protocol_buffer_message_name>,subject=<schema_registry_subject>` to read the latest schema from a subject in the schema registry, which allows one to use Iceberg topics with protobuf and not having to encode the messages using the Schema Registry wire format.

rockwotj committed:
datalake/manager: remove unused member

rockwotj committed:
iceberg: support latest_protobuf_value mode

A new iceberg mode that also requires additional metadata (which
protobuf message to use), which means we no longer can store
iceberg_mode as a simple enum.

I converted it to a struct that is very similar API wise to the enum,
but for the protobuf case the API diverges.

This is a minimal change to add the new variant and update the API for
iceberg_mode. There are likely changes were we copy iceberg mode around,
and we might want to move, although TBH I expect the protobuf message
name to fit in the SSO optimization or be < 32 characters so I am not
sure the moves will help at all, but maybe references can be used
instead. I'll investigate updating that in another PR, although I expect
it to not be copied that much as this is really only used in setup for
the datalake and in admin-like requests.

rockwotj committed:
datalake/sr: add a schema resolver based on latest subject version

Implements the proposal to work with protobuf messages that don't use
the sr wire format. Instead we trust the latest protobuf is
backwards compatible on the wire, then parse the record based on that.

We always assume the topic naming strategy. Right now there is no
caching of the schema, but we will add support for this in a followup
commit.

rockwotj committed:
pp/sr: add a method to lookup a schema by full name

To be used in datalake for alternative iceberg_modes

rockwotj committed:
datalake: hookup latest_protobuf_value

rockwotj committed:
dt/datalake: add test for latest_protobuf_value mode

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--1ddf9a3f-d58b-42c6-b40e-c62989d21e33-->
<!--TEST RESULTS-->
<!--51e28191-afbc-4c97-9f16-bb69c366576b-->
<!--TEST RESULTS-->
<!--e0b28c2f-a2fc-4a7d-b21c-03bc5a2b6787-->
<!--TEST RESULTS-->
<!--0e4b6267-2795-4e32-bce4-f70d1a9a134e-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/63081">build#63081</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.rest_catalog_connection_test.RestCatalogConnectionTest.test_redpanda_connection_to_rest_catalog.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/63081#0195916f-3073-4e1e-957f-54082ade1b75 | FLAKY | 1/2 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/63081#0195916f-3073-4e1e-957f-54082ade1b75 | FLAKY | 1/2 |
| rptest.tests.write_caching_fi_e2e_test.WriteCachingFailureInjectionE2ETest.test_crash_all_with_consumer_group | ducktape | https://buildkite.com/redpanda/redpanda/builds/63081#01959155-c278-49c1-b2ab-9e8db0cd91a5 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/63180">build#63180</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:----------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| commands_serialization_test_rpunit.commands_serialization_test_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/63180#01959686-299a-4a6c-84f6-9773c9b87637 | FAIL | 0/2 |
| commands_serialization_test_rpunit.commands_serialization_test_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/63180#01959686-299b-40d5-9547-6515b70bd1b9 | FAIL | 0/2 |
| partition_balancer_planner_test_rpunit.partition_balancer_planner_test_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/63180#01959686-299a-4a6c-84f6-9773c9b87637 | FAIL | 0/2 |
| partition_balancer_planner_test_rpunit.partition_balancer_planner_test_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/63180#01959686-299b-40d5-9547-6515b70bd1b9 | FAIL | 0/2 |
| partition_balancer_simulator_test_rpunit.partition_balancer_simulator_test_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/63180#01959686-299a-4a6c-84f6-9773c9b87637 | FAIL | 0/2 |
| partition_balancer_simulator_test_rpunit.partition_balancer_simulator_test_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/63180#01959686-299b-40d5-9547-6515b70bd1b9 | FAIL | 0/2 |
| serialization_rt_test_rpunit.serialization_rt_test_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/63180#01959686-299a-4a6c-84f6-9773c9b87637 | FAIL | 0/2 |
| serialization_rt_test_rpunit.serialization_rt_test_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/63180#01959686-299b-40d5-9547-6515b70bd1b9 | FAIL | 0/2 |
| test_compat_rpunit.test_compat_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/63180#01959686-299a-4a6c-84f6-9773c9b87637 | FAIL | 0/2 |
| test_compat_rpunit.test_compat_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/63180#01959686-299b-40d5-9547-6515b70bd1b9 | FAIL | 0/2 |
| test_group_tx_compaction_rpunit.test_group_tx_compaction_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/63180#01959686-299a-4a6c-84f6-9773c9b87637 | FAIL | 0/2 |
| test_group_tx_compaction_rpunit.test_group_tx_compaction_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/63180#01959686-299b-40d5-9547-6515b70bd1b9 | FAIL | 0/2 |
| topic_manifest_rpunit.topic_manifest_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/63180#01959686-299a-4a6c-84f6-9773c9b87637 | FAIL | 0/2 |
| topic_manifest_rpunit.topic_manifest_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/63180#01959686-299b-40d5-9547-6515b70bd1b9 | FAIL | 0/2 |
| topic_updates_dispatcher_test_rpunit.topic_updates_dispatcher_test_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/63180#01959686-299a-4a6c-84f6-9773c9b87637 | FAIL | 0/2 |
| topic_updates_dispatcher_test_rpunit.topic_updates_dispatcher_test_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/63180#01959686-299b-40d5-9547-6515b70bd1b9 | FAIL | 0/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/63197">build#63197</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/63197#01959a4d-fc90-449b-a0a9-6e1ec6efa85b | FLAKY | 1/2 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.True | ducktape | https://buildkite.com/redpanda/redpanda/builds/63197#01959a4d-fc90-48f1-b04e-3af54ed7168d | FLAKY | 1/3 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/63197#01959a50-08ce-4ae9-aa2d-780f84711259 | FLAKY | 1/5 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/63625">build#63625</a></summary>


| test_id                                                                                               | test_kind   | job_url                                                                                   | test_status   | passed   |
|:------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape    | https://buildkite.com/redpanda/redpanda/builds/63625#0195cb29-b133-49e6-88ae-9538c1b30959 | FLAKY         | 1/2      |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_mount_inexistent                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/63625#0195cb29-b136-4de3-b64a-16d794054d11 | FLAKY         | 1/2      |
</details>
<!--/body-->

nvartolomei commented on src/v/model/model.cc:
```diff
@@ -582,27 +584,74 @@ std::istream& operator>>(std::istream& is, recovery_validation_mode& vm) {
+iceberg_mode iceberg_mode::value_schema_id_prefix
+  = iceberg_mode::make<iceberg_mode::variant::value_schema_id_prefix>();
+
+void write(iobuf& out, const iceberg_mode& m) {
```
offline log viewer needs to be updated too https://github.com/redpanda-data/redpanda/blob/4c87a134d3896d806e0d817d97957590170dae75/tools/offline_log_viewer/controller.py#L141

rockwotj replied:
Nice catch - done!

nvartolomei commented on src/v/model/model.cc:
```diff
@@ -582,27 +584,74 @@ std::istream& operator>>(std::istream& is, recovery_validation_mode& vm) {
+        read_nested(in, s, bytes_left_limit);
+        m = iceberg_mode::latest_protobuf_value(s);
+        break;
+    }
```
wouldn't it be better to return inside the switch statement and if we get there to throw an exception? in the unlikely case we mess up somwhere/forget a feature gate we would end up with disabled iceberg.

this is backwards compatible with the old code I suppose?

rockwotj replied:
I think throwing is a good idea, will do.

rockwotj replied:
done

andrwng commented on src/v/datalake/record_schema_resolver.cc:
```diff
@@ -330,4 +334,136 @@ record_schema_resolver::resolve_identifier(schema_identifier ident) const {
+ss::future<checked<type_and_buf, type_resolver::errc>>
+latest_protobuf_schema_resolver::resolve_buf_type(
+  std::optional<iobuf> b) const {
+    // Prevent multiple fibers from resolving the protobuf. Since the subject
+    // is fixed there would be no harm in letting them race either, but it seems
+    // better to not waste work.
```
FWIW each resolver on the data path is only used by a single translator, so there _shouldn't_ be these kinds of races unless we begin to share the cache.

There is a shared schema cache, but it's the S3 fifo cache, not sure it lends itself well to having a TTL for schemas

rockwotj replied:
Thanks, removed

andrwng reviewed (dismissed):
Modulo making sure we add more tests (e.g. caching/TTL, backwards compat for protobuf, etc) these changes pretty much lgtm

rockwotj committed:
dl/schema: remove mutex

This is only used by a single fiber

rockwotj committed:
datalake: support configurable latest schema cache duration

rockwotj commented:
Force push: added cluster config for schema cache duration, and added tests for schema evolution + backwards compat

andrwng reviewed (dismissed):
Thanks for adding to the test! This is 🔥

piyushredpanda commented:
Good to merge, @rockwotj ?

piyushredpanda commented:
cc: @Feediver1 and @mattschumpert

rockwotj commented:
@mattschumpert wanted to talk configuration naming and semantics when I am back in office. I think we just land it in a later rc before the full cut?

rockwotj committed:
schema: fix posting identical schemas

They should get the same ID but be available under multiple subjects.

rockwotj committed:
datalake: use latest subject iceberg mode

Instead of a specific protobuf one, we'll just have a generic schema
mode for latest schema within a subject.

By default we'll use the topic naming strategy to determine the subject
and we'll assume the first protobuf message in a file. Otherwise it's
possible for a user to override these values using inline key/value
pairs to the topic config.

rockwotj commented:
Okay I've gone ahead and pushed two new commits one to enable some unit tests and another that changes the behavior to be like the outcome of my meeting with @mattschumpert

rockwotj commented:
cc: @kbatuigas as a docs FYI

andrwng commented on src/v/datalake/record_schema_resolver.cc:
```diff
@@ -404,48 +396,45 @@ latest_protobuf_schema_resolver::resolve_buf_type(
-        -> checked<ppsr::protobuf_schema_definition, type_resolver::errc> {
+      [&latest_schema, &shared_schema](const ppsr::avro_schema_definition& def)
+        -> checked<resolved_type, type_resolver::errc> {
+          return translate_avro_schema(
+            def, latest_schema.id, std::move(shared_schema));
```
Oh Avro support is surprising from the CL, but great! I think the release note needs an update.

andrwng replied:
>which allows one to use Iceberg topics with protobuf and not having to encode the messages using the Schema Registry wire format.

Specifically this part that calls out protobuf specifically, given it's true for Avro too now

rockwotj replied:
I actually don't want to encourage avro usage at all, because it fundamentally does not allow schema evolution. Not supporting it explicitly seems a bit harsh and @mattschumpert was for allowing it. Schema evolution in avro works by having the new and old schemas and evolving the parsed data. Which means in this mode we'll fail to parse old avro records (because we only ever have one schema). I will make sure we have a note on the docs this is not recommended to use avro with this mode, but having the mode based on how we lookup the schema is better than something specific for protobuf and is easier to explain.

andrwng replied:
Yea fair. It does seem like a decent addition regardless, especially if users know there won't be schema changes. It'd be interesting to understand the set of customers using Avro without the SR format to see if there is some other mode we ought to be supporting

rockwotj replied:
Yeah agreed. I think this no one does this (schema registry was initially created only for avro and protobuf then json came later). If they do they have to be communicating the schema out of band somehow (maybe via a header?) and I am interested to know if there are existing patterns out there.

rockwotj commented:
Force push: fix rputil build

rockwotj commented:
/backport v25.1.x
